modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/RoutersExecutionTestCase.java

```diff
@@ -47,6 +58,60 @@ public void singleRouteRouter() throws Exception {
     assertThat(internalEvent.getVariables().get("newAttributes").getValue(), is("other"));
   }
 
+  /**
+   * Executes once a router that internally calls concurrently the same Chain
+   */
+  @Test
+  public void concurrentRouter() throws Exception {
+    CoreEvent internalEvent = flowRunner("concurrentRouteExecutor").run();
+
+    assertThat(internalEvent.getMessage().getPayload().getValue(), is("SUCCESS"));
+  }
+
+  /**
+   * Executes the same flow concurrently to check that no race condition exists because
+   * two different instances of Chain are being used
+   */
+  @Test
+  public void concurrentRouterExecution() throws Exception {
+
+    final ExecutorService executor = newFixedThreadPool(2);
+
+    final Latch beginLatch = new Latch();
+    final CountDownLatch assertLatch = new CountDownLatch(2);
+    final Consumer<Reference<CoreEvent>> runner = reference -> {
+      try {
+        beginLatch.await();
+        reference.set(flowRunner("singleRouteRouter")
+            .withPayload("CustomPayload")
+            .run());
+        assertLatch.countDown();
+      } catch (Exception e) {
+        e.printStackTrace();
+        fail(e.getMessage());
+      }
+    };
+
+    final Reference<CoreEvent> first = new Reference<>();
+    final Reference<CoreEvent> second = new Reference<>();
+
+    executor.submit(() -> runner.accept(first));
+    executor.submit(() -> runner.accept(second));
+
+    beginLatch.release();
+    assertLatch.await();
```

Comment:
timeout

Reply:
we have the test timeout already, don't we?

Reply:
no need to wait a whole minute for this to fail